test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs

```diff
@@ -846,5 +848,86 @@ public async Task Uninstall_WithMultiplePackagesThatDependOnEachOther_PackageGoe
             uiwindow.AssertPackageNameAndType(TestPackageName, NuGet.VisualStudio.PackageLevel.TopLevel);
             uiwindow.AssertPackageNameAndType(transitivePackageName, NuGet.VisualStudio.PackageLevel.Transitive);
         }
+
+        [TestMethod]
+        [DataRow(ProjectTemplate.NetCoreConsoleApp)]
+        [DataRow(ProjectTemplate.ConsoleApplication)]
+        [Timeout(DefaultTimeout)]
+        public async Task VerifyRestorePackageByRestoreNuGetPackages(ProjectTemplate projectTemplate)
+        {
+            // Arrange
+            string packageFolderPath;
+            string installedPackageFolderPath;
+
+            await CommonUtility.CreatePackageInSourceAsync(_pathContext.PackageSource, TestPackageName, TestPackageVersionV1);
+
+            NuGetApexTestService nugetTestService = GetNuGetTestService();
+
+            var solutionService = VisualStudio.Get<SolutionService>();
+            solutionService.CreateEmptySolution("TestSolution", _pathContext.SolutionRoot);
+            var project = solutionService.AddProject(ProjectLanguage.CSharp, projectTemplate, "TestProject");
+            VisualStudio.ClearOutputWindow();
+            solutionService.SaveAll();
+
+            CommonUtility.OpenNuGetPackageManagerWithDte(VisualStudio, Logger);
+
+            var uiwindow = nugetTestService.GetUIWindowfromProject(project);
+            uiwindow.InstallPackageFromUI(TestPackageName, TestPackageVersionV1);
+            solutionService.Build();
+
+            if (projectTemplate.Equals(ProjectTemplate.NetCoreConsoleApp))
+            {
+                packageFolderPath = _pathContext.UserPackagesFolder;
+                installedPackageFolderPath = Path.Combine(packageFolderPath, TestPackageName);
+            }
+            else
+            {
+                packageFolderPath = _pathContext.PackagesV2;
+                installedPackageFolderPath = Path.Combine(packageFolderPath, "Contoso.A.1.0.0");
+            }
+
+            Directory.Exists(installedPackageFolderPath).Should().BeTrue();
+
+            // Act
+            Directory.Delete(packageFolderPath, true);
+            Directory.Exists(packageFolderPath).Should().BeFalse();
+            CommonUtility.RestoreNuGetPackages(VisualStudio, Logger);
+
+            // Assert (It takes some time for the package folder to be back after the solution is restored.)
+            Thread.Sleep(200);
```

Comment:
I think this is something that has the potential to be flaky. 
Is 200ms enough for the restore to complete?
What if we end up on a slower machine, or we hit a long GC or something?

Reply:
Sorry, I didn't take this scenario into consideration, thanks for your reminder. I've updated this method to use the `WaitForDirectoryExists()` method to assert restore after the packages folder back and set the maximum wait time to 10 seconds.